libs/lib-postgres/src/db/connection/DatabaseClient.ts

```diff
@@ -0,0 +1,250 @@
+import * as lib_postgres from '@powersync/lib-service-postgres';
+import * as pgwire from '@powersync/service-jpgwire';
+import pDefer, { DeferredPromise } from 'p-defer';
+import { AbstractPostgresConnection, sql } from './AbstractPostgresConnection.js';
+import { ConnectionLease, ConnectionSlot, NotificationListener } from './ConnectionSlot.js';
+import { WrappedConnection } from './WrappedConnection.js';
+
+export type DatabaseClientOptions = {
+  config: lib_postgres.NormalizedBasePostgresConnectionConfig;
+  /**
+   * Optional schema which will be used as the default search path
+   */
+  schema?: string;
+  /**
+   * Notification channels to listen to.
+   */
+  notificationChannels?: string[];
+};
+
+export type DatabaseClientListener = NotificationListener & {
+  connectionCreated?: (connection: pgwire.PgConnection) => Promise<void>;
+};
+
+export const TRANSACTION_CONNECTION_COUNT = 5;
+
+export class DatabaseClient extends AbstractPostgresConnection<DatabaseClientListener> {
+  closed: boolean;
+
+  protected pool: pgwire.PgClient;
+  protected connections: ConnectionSlot[];
```

Comment:
This appears to use a combination of a pgwire pool, and a custom connection pool. Where should one be used versus the other?

Reply:
The connection slots are currently used for providing transaction and notification functionality which I believe is not possible with the pgwire pool. The pgwire connection pool is used initial for basic individual statements, but it could be replaced with the custom pool if it's overkill to have 2 pools.